src/plugins/data/public/query/state_sync/sync_filters.ts

```diff
@@ -0,0 +1,167 @@
+  timeFilter: TimefilterContract
+) => {
+  const defaultState: FiltersSyncState = {
+    time: timeFilter.getTime(),
+    refreshInterval: timeFilter.getRefreshInterval(),
+    filters: filterManager.getGlobalFilters(),
+  };
+
+  const initialStateFromUrl = urlStateStorage.get<FiltersSyncState>(GLOBAL_STATE_STORAGE_KEY);
+
+  // remember whether there were info in the URL
+  const hasInheritedFiltersFromUrl = Boolean(
+    initialStateFromUrl && Object.keys(initialStateFromUrl).length
+  );
+
+  const initialState: FiltersSyncState = {
+    ...defaultState,
+    ...initialStateFromUrl,
+  };
+
+  const filtersSyncStateContainer = createStateContainer(
+    initialState,
+    {
+      setTime: (state: FiltersSyncState) => (time: TimeRange) => ({ ...state, time }),
+      setRefreshInterval: (state: FiltersSyncState) => (refreshInterval: RefreshInterval) => ({
+        ...state,
+        refreshInterval,
+      }),
+      setFilters: (state: FiltersSyncState) => (filters: esFilters.Filter[]) => ({
+        ...state,
+        filters,
+      }),
+    },
+    {
+      time: (state: FiltersSyncState) => () => state.time,
+      refreshInterval: (state: FiltersSyncState) => () => state.refreshInterval,
+      filters: (state: FiltersSyncState) => () => state.filters,
+    }
+  );
+
+  const subs: Subscription[] = [
+    timeFilter.getTimeUpdate$().subscribe(() => {
+      filtersSyncStateContainer.transitions.setTime(timeFilter.getTime());
+    }),
+    timeFilter.getRefreshIntervalUpdate$().subscribe(() => {
+      filtersSyncStateContainer.transitions.setRefreshInterval(timeFilter.getRefreshInterval());
+    }),
+    filterManager
+      .getUpdates$()
+      .pipe(
+        map(() => filterManager.getGlobalFilters()),
+        filter(newGlobalFilters => {
+          // continue only if global filters changed
+          // and ignore app state filters
+          const oldGlobalFilters = filtersSyncStateContainer.get().filters;
+          return (
+            !oldGlobalFilters ||
+            !compareFilters(newGlobalFilters, oldGlobalFilters, COMPARE_ALL_OPTIONS)
+          );
+        })
+      )
+      .subscribe(newGlobalFilters => {
+        filtersSyncStateContainer.transitions.setFilters(newGlobalFilters);
+      }),
+    filtersSyncStateContainer.state$.subscribe(
+      ({ time, filters: globalFilters, refreshInterval }) => {
+        if (time && !_.isEqual(time, timeFilter.getTime())) {
+          timeFilter.setTime(_.cloneDeep(time));
+        }
+
+        if (refreshInterval && !_.isEqual(refreshInterval, timeFilter.getRefreshInterval())) {
+          timeFilter.setRefreshInterval(_.cloneDeep(refreshInterval));
+        }
+
+        if (
+          globalFilters &&
+          !compareFilters(globalFilters, filterManager.getGlobalFilters(), COMPARE_ALL_OPTIONS)
+        ) {
+          globalFilters = _.cloneDeep(globalFilters);
+          FilterManager.setFiltersStore(globalFilters, esFilters.FilterStateStore.GLOBAL_STATE);
```

Comment:
nit: That reads super strange (both `filterManager` and `FilterManager`) - is there a reason this is a static method on `FilterManager` instead of a separately exported helper function like `compareFilters`? (same in `src/legacy/core_plugins/kibana/public/dashboard/np_ready/lib/sync_app_filters.ts`)

Reply:
cc @lizozom

Reply:
No strong feelings either way.
Don't mind if this is a util.

Reply:
Not relevant anymore because of this change: https://github.com/elastic/kibana/pull/55158/files#r370079408